### ISSUE TITLE
Can't transpose np.where in custom_jvp for det

### ISSUE THREAD

**Issue by dpfau** [google/jax#2756]
This is mostly meant as a reminder for Matt.

I am working on a new jvp rule for linalg.det to address #2380. In my jvp computation, I mask out any NaNs due to nonsingular matrices and replace them with zeros (this is correct for any matrix with rank less than n-1, while the rank n-1 case is handled separately). However, there is no transpose rule for the select used to do this masking. A workaround is needed.

Offending line is here: https://github.com/dpfau/jax/blob/master/jax/numpy/linalg.py#L207

**Comment by shoyer** [comment]
In theory, select has a transpose rule. I would be curious to see a self-contained example of what goes wrong here.

**Comment by dpfau** [comment]
I guess the description is a little misleading. If you run `linalg_test.py` in my fork (I need to check but you may have to fiddle with some bits), it will fail in `_select_transpose_rule` in `lax.py` at the line `assert not ad.is_undefined_primal(pred)`, and when you check what `pred` is, it's a `select`. If you remove the `np.where` and just return `x` from `cofactor_solve` then the tests with full-rank matrices pass fine, but it still returns NaN for low rank matrices.

**Comment by mattjj** [comment]
Hrm NumpyLinalgTest.testDetOfSingularMatrix passes for me when I just use the `custom_jvp` rule, or just use the `custom_vjp` rule. Can you say more about how to repro this?

**Comment by mattjj** [comment]
The problem for the JVP rule is that in the expression `np.where(np.isnan(x), np.zeros_like(x), x)` the subexpression `np.isnan(x)` isn't linear in `x`. As a result, those values aren't even available when performing reverse-mode differentiation on this function.

It seems like we should be able to figure out where the invalid values are just based on the matrix `a` or its LU decomposition `lu`...

**Comment by mattjj** [comment]
This is dumb, but as a proof of concept if you replace the `return` statement in `_cofactor_solve` with this:

```python
  d = np.ones(x.shape, x.dtype)  # d stands for "dummy"
  d = lax_linalg.triangular_solve(lu, d, left_side=True, lower=True,
                                  unit_diagonal=True)
  d = np.concatenate((d[..., :-1, :] * partial_det[..., -1, None, None],
                      d[..., -1:, :]), axis=-2)
  d = lax_linalg.triangular_solve(lu, d, left_side=True, lower=False)

  return partial_det[..., -1], np.where(np.isnan(d), np.zeros_like(x), x)
```

Then the test passes! Woo!

There's probably a more efficient way to extract the singularity information we need. But we can't run the solve on `b` (aka `x`) to do it.

WDYT?

**Comment by dpfau** [comment]
Yeah, this works, but I'm worried that it takes twice the computation. I suppose asymptotically, the LU decomposition will take up most of the time, but triangular_solve still isn't quick. But I don't have a better idea for reliably checking for singularity for all backends and precisions.

Also - you should be able to drop everything but the last triangular_solve. The first triangular_solve will never be singular because `unit_diagonal=True`. So try

```
  d = np.ones(x.shape, x.dtype)  # d stands for "dummy"
  d = lax_linalg.triangular_solve(lu, d, left_side=True, lower=False)

  return partial_det[..., -1], np.where(np.isnan(d), np.zeros_like(x), x)
```
and let me know if it works?

**Comment by dpfau** [comment]
Also you probably don't need `d` to have exactly the shape of `x`. Try:

```
  d = np.ones((x.shape[0], 1), x.dtype)  # d stands for "dummy"
  d = lax_linalg.triangular_solve(lu, d, left_side=True, lower=False)

  return partial_det[..., -1], np.where(np.isnan(d), np.zeros_like(x), x)
```

Actually, I'll give it a try...

**Comment by dpfau** [comment]
OK, forgot to handle some shaping issues in the solution above, and it turns out not *all* entries in `d` will be NaN if the matrix is singular. This passes:

```
  d = np.ones(x.shape[:-1], x.dtype)  # d stands for "dummy"
  d = lax_linalg.triangular_solve(lu, d, left_side=True, lower=False)
  d = np.any(np.logical_or(np.isnan(d), np.isinf(d)), axis=-1)
  d = np.tile(d[..., None], d.shape + x.shape[-2:])
  return partial_det[..., -1], np.where(d, np.zeros_like(x), x)
```

**Comment by dpfau** [comment]
Although it only seems to pass in forward mode. Looking into reverse mode.

**Comment by dpfau** [comment]
To be honest, I'm finding it a lot harder to debug the reverse mode since I can't just put `print` or `pdb.set_trace()` everywhere. Matt, could you take a look and figure out why your solution works and mine doesn't?

**Comment by mattjj** [comment]
Yes, will take a look! Thanks for the progress on this. (And good point about unit_diagonal!)

**Comment by mattjj** [comment]
Picking this back up. The reverse mode effect could be related to my long comment on #1052. Will investigate.

**Comment by mattjj** [comment]
Yup that was it. After the `iotas = ...` line I wrote this:

```python
  d = np.ones(x.shape[:-1], x.dtype)  # d stands for "dude, cool"
  d = lax_linalg.triangular_solve(lu, d, left_side=True, lower=False)
  d = np.any(np.logical_or(np.isnan(d), np.isinf(d)), axis=-1)
  d = np.tile(d[..., None, None], d.ndim*(1,) + x.shape[-2:])

  x = np.where(d, np.zeros_like(x), x)
  x = x[iotas[:-1] + (permutation, slice(None))]
  x = lax_linalg.triangular_solve(lu, x, left_side=True, lower=True,
                                  unit_diagonal=True)
  x = np.concatenate((x[..., :-1, :] * partial_det[..., -1, None, None],
                      x[..., -1:, :]), axis=-2)
  x = lax_linalg.triangular_solve(lu, x, left_side=True, lower=False)
  x = np.where(d, np.zeros_like(x), x)

  return partial_det[..., -1], x
```

And got this from the test you suggested:

```
In [1]: run pfauuu
[[0. 0.]
 [0. 0.]]
```

**Comment by dpfau** [comment]
Ah, interesting...

**Comment by mattjj** [comment]
Maybe we could write `d = np.ones((x.shape[-2], 1), x.dtype)` and rely on the broadcasting behavior of `np.where`.

**Comment by dpfau** [comment]
OK, now it's failing if I increase the order of the test to 2. Also realized something...at the moment, this should zero out rank n-1 matrices as well, which is actually incorrect...in fact the reason for adding all this complex machinery is so that it handles the n-1 case correctly. Yet that's not showing up in the test for some reason...will investigate tomorrow.

**Comment by dpfau** [comment]
So after some investigation, it seems like it works fine for a rank n-1 matrix which the current JAX version gives NaN for. I've included it in the tests.

Next - why does the second-order derivative fail for low-rank matrices?

**Comment by dpfau** [comment]
Fixed and submitted as PR #2809
